pkg/cmd/admin/image/verify-signature.go

```diff
@@ -0,0 +1,368 @@
+	ref       sigtypes.ImageReference
+	manifest  []byte
+	signature []byte
+}
+
+func newUnparsedImage(expectedIdentity string, img *imageapi.Image, signature []byte) (sigtypes.UnparsedImage, error) {
+	ref, err := parseDockerReference("//" + expectedIdentity)
+	if err != nil {
+		return nil, fmt.Errorf("Invalid --expected-identity: %v", err)
+	}
+	return &unparsedImage{ref: ref, manifest: []byte(img.DockerImageManifest), signature: signature}, nil
+}
+
+// Reference returns the reference used to set up this source, _as specified by the user_
+// (not as the image itself, or its underlying storage, claims).  This can be used e.g. to determine which public keys are trusted for this image.
+func (ui *unparsedImage) Reference() sigtypes.ImageReference {
+	return ui.ref
+}
+
+// Close removes resources associated with an initialized UnparsedImage, if any.
+func (ui *unparsedImage) Close() error {
+	return nil
+}
+
+// Manifest is like ImageSource.GetManifest, but the result is cached; it is OK to call this however often you need.
+func (ui *unparsedImage) Manifest() ([]byte, string, error) {
+	return ui.manifest, "", nil
+}
+
+// Signatures is like ImageSource.GetSignatures, but the result is cached; it is OK to call this however often you need.
+func (ui *unparsedImage) Signatures() ([][]byte, error) {
+	return [][]byte{ui.signature}, nil
+}
+
+func (o *VerifyImageSignatureOptions) Run() error {
+	img, err := o.Client.Images().Get(o.InputImage)
+	if err != nil {
+		return err
+	}
+	if len(img.Signatures) == 0 {
+		return fmt.Errorf("%s does not have any signature", img.Name)
+	}
+
+	policy, err := signature.DefaultPolicy(nil)
+	if err != nil {
+		return fmt.Errorf("Error reading signature verification policy: %v", err)
+	}
+	pc, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return fmt.Errorf("Error preparing for signature verification: %v", err)
+	}
+	defer pc.Destroy()
+
+	for i, s := range img.Signatures {
+		// If --remove is specified, just erase the existing signature verification for all
+		// signatures.
+		// TODO: This should probably need to handle removal of a single signature.
+		if o.Remove {
+			o.clearSignatureVerificationStatus(&img.Signatures[i])
+			continue
+		}
+
+		// Verify the signature against the policy
+		signedBy, err := o.verifySignature(pc, img, s.Content)
+		if err != nil {
+			fmt.Fprintf(o.ErrOut, "error verifying %s signature %d: %v\n", o.InputImage, i, err)
+			o.clearSignatureVerificationStatus(&img.Signatures[i])
+			continue
+		}
+		fmt.Fprintf(o.Out, "%s signature %d is verified (signed by key: %q)\n", o.InputImage, i, signedBy)
+
+		now := unversioned.Now()
+		newConditions := []imageapi.SignatureCondition{
+			{
+				Type:               imageapi.SignatureTrusted,
+				Status:             kapi.ConditionTrue,
+				LastProbeTime:      now,
+				LastTransitionTime: now,
+				Reason:             "verified manually",
+				Message:            fmt.Sprintf("verified by user %s", o.CurrentUser),
```

Comment:
also records quotes into API object ;-) i don't mind honestly, but I found no quotes nicer as they don't have to be escaped.